src/crimson/os/seastore/journal.cc

```diff
@@ -50,163 +49,22 @@ segment_nonce_t generate_nonce(
-{
-  ceph::bufferlist to_write = encode_record(
-    rsize, std::move(record), segment_manager.get_block_size(),
-    committed_to, current_segment_nonce);
-  auto target = written_to;
-  assert((to_write.length() % segment_manager.get_block_size()) == 0);
-  written_to += to_write.length();
-  logger().debug(
-    "write_record, mdlength {}, dlength {}, target {}",
-    rsize.mdlength,
-    rsize.dlength,
-    target);
-
-  auto segment_id = current_journal_segment->get_segment_id();
-
-  // Start write under the current exclusive stage, but wait for it
-  // in the device_submission concurrent stage to permit multiple
-  // overlapping writes.
-  auto write_fut = current_journal_segment->write(target, to_write);
-  return handle.enter(write_pipeline->device_submission
-  ).then([write_fut = std::move(write_fut)]() mutable {
-    return std::move(write_fut
-    ).handle_error(
-      write_record_ertr::pass_further{},
-      crimson::ct_error::assert_all{
-	"Invalid error in Journal::write_record"
-      }
-    );
-  }).safe_then([this, &handle] {
-    return handle.enter(write_pipeline->finalize);
-  }).safe_then([this, target, segment_id] {
-    logger().debug(
-      "write_record: commit target {}",
-      target);
-    if (segment_id == current_journal_segment->get_segment_id()) {
-      assert(committed_to < target);
-      committed_to = target;
-    }
-    return write_record_ret(
-      write_record_ertr::ready_future_marker{},
-      paddr_t{
-	segment_id,
-	target});
-  });
-}
-
-bool Journal::needs_roll(segment_off_t length) const
-{
-  return length + written_to >
-    current_journal_segment->get_write_capacity();
-}
-
-Journal::roll_journal_segment_ertr::future<segment_seq_t>
-Journal::roll_journal_segment()
-{
-  auto old_segment_id = current_journal_segment ?
-    current_journal_segment->get_segment_id() :
-    NULL_SEG_ID;
-
-  return (current_journal_segment ?
-	  current_journal_segment->close() :
-	  Segment::close_ertr::now()).safe_then([this] {
-      return segment_provider->get_segment(segment_manager.get_device_id());
-    }).safe_then([this](auto segment) {
-      return segment_manager.open(segment);
-    }).safe_then([this](auto sref) {
-      current_journal_segment = sref;
-      written_to = 0;
-      return initialize_segment(*current_journal_segment);
-    }).safe_then([=](auto seq) {
-      if (old_segment_id != NULL_SEG_ID) {
-	segment_provider->close_segment(old_segment_id);
-      }
-      segment_provider->set_journal_segment(
-	current_journal_segment->get_segment_id(),
-	seq);
-      return seq;
-    }).handle_error(
-      roll_journal_segment_ertr::pass_further{},
-      crimson::ct_error::all_same_way([] { ceph_assert(0 == "TODO"); })
-    );
-}
-
-Journal::open_for_write_ret Journal::open_for_write()
+Journal::Journal(
+  SegmentManager& segment_manager,
+  ExtentReader& scanner)
+  : journal_segment_manager(segment_manager),
+    record_submitter(2, journal_segment_manager),
```

Comment:
Should there be a config for io_depth_limit?

Reply:
Yeah, I think so.

Reply:
added options